cloudinit/sources/DataSourceCloudStack.py

```diff
@@ -256,7 +256,17 @@ def get_vr_address():
                 if len(words) > 2:
                     dhcptok = words[2]
                     LOG.debug("Found DHCP identifier %s", dhcptok)
-                    latest_address = dhcptok
+                    return dhcptok
+
+    # Try NetworkManager lease files format next...
+    with open(lease_file, "r") as fd:
+        for line in fd:
+            if "SERVER_ADDRESS" in line:
```

Comment:
It may be possible move this if clause into an elif in the previous parsing of the lease file.

```
% git diff cloudinit/sources/DataSourceCloudStack.py
diff --git a/cloudinit/sources/DataSourceCloudStack.py b/cloudinit/sources/DataSourceCloudStack.py
index f185dc7..838996d 100644
--- a/cloudinit/sources/DataSourceCloudStack.py
+++ b/cloudinit/sources/DataSourceCloudStack.py
@@ -239,6 +239,14 @@ def get_vr_address():
                     dhcptok = words[2]
                     LOG.debug("Found DHCP identifier %s", dhcptok)
                     latest_address = dhcptok
+            # Try NetworkManager lease file format
+            elif "SERVER_ADDRESS" in line:
+                 words = line.strip().split("=")
+                 if len(words) == 2:
+                     dhcptok = words[1]
+                     LOG.debug("Found DHCP identifier %s", dhcptok)
+                    latest_address = dhcptok
+

```
Also, you changed the logic for 'dhcp-server-identify' to return at the *first* match, where previously, it would parse the entire file and return the last (you removed the latest_address = dhcptok assignment).

It's not clear to me if that will affect things since the lease format isn't documented here.  Can you confirm that there will be only one entry per lease file; if not then we should continue to set the latest_address to the found dhcptok and complete the parsing of the lease file.

Reply:
I applied suggestion ,but it seems the issue is already resolved in different way: https://bugs.launchpad.net/cloud-init/+bug/1843334/comments/8.